src/IssueViz.Security/Taint/TaintStore.cs

```diff
@@ -39,21 +40,85 @@ internal interface ITaintStore : IIssuesStore
         /// Returns additional analysis information for the existing visualizations in the store.
         /// </summary>
         AnalysisInformation GetAnalysisInformation();
+
+        /// <summary>
+        /// Add the given issue to the existing list of visualizations.
+        /// If <see cref="GetAnalysisInformation"/> is null, the operation is ignored.
+        /// </summary>
+        void Add(IAnalysisIssueVisualization issueVisualization);
+
+        /// <summary>
+        /// Removes an issue with the given key from the existing list of visualizations.
+        /// If no matching issue is found, the operation is ignored.
+        /// </summary>
+        void Remove(string issueKey);
     }
 
     [Export(typeof(ITaintStore))]
     [Export(typeof(IIssuesStore))]
     [PartCreationPolicy(CreationPolicy.Shared)]
     internal sealed class TaintStore : ITaintStore
     {
-        private IAnalysisIssueVisualization[] taintVulnerabilities = Array.Empty<IAnalysisIssueVisualization>();
+        public event EventHandler<IssuesChangedEventArgs> IssuesChanged;
+
+        private static readonly object Locker = new object();
+
+        private List<IAnalysisIssueVisualization> taintVulnerabilities = Array.Empty<IAnalysisIssueVisualization>().ToList();
         private AnalysisInformation analysisInformation;
 
-        public IReadOnlyCollection<IAnalysisIssueVisualization> GetAll() => taintVulnerabilities;
-        
+        public IReadOnlyCollection<IAnalysisIssueVisualization> GetAll()
+        {
+            lock (Locker)
+            {
+                return taintVulnerabilities;
+            }
+        }
+
         public AnalysisInformation GetAnalysisInformation() => analysisInformation;
 
-        public event EventHandler<IssuesChangedEventArgs> IssuesChanged;
+        public void Add(IAnalysisIssueVisualization issueVisualization)
```

Comment:
Don't we get the taint event notifications in batch? If so, do we want to update the store in batch, or one item at a time?

Batch would be more efficient, but one at a time is probably ok for now. It might trigger multiple UI updates if the taint tool window is visible,  but the number of events should be relatively small so it should cause a problem.

Reply:
Unfortunately batch is only for issues. Taint is one by one. Consistency...